src/algebra/commute.lean

```diff
@@ -0,0 +1,371 @@
+eq.symm h
+
+protected theorem symm_iff {a b : S} : commute a b ↔ commute b a := ⟨commute.symm, commute.symm⟩
+
+end commute
+
+-- Definitions and trivial theorems about them
+section defs
+
+variables {S : Type*} [has_mul S]
+
+/-- Centralizer of an element `a : S` is the set of elements that commute with `a`. -/
+def centralizer (a : S) : set S := { x | commute a x }
+
+theorem mem_centralizer {a b : S} : b ∈ centralizer a ↔ commute a b := iff.rfl
+
+/-- Centralizer of a set `T` is the set of elements that commute with all `a ∈ T`. -/
+protected def set.centralizer (s : set S) : set S := { x | ∀ a ∈ s, commute a x }
+
+protected theorem set.mem_centralizer (s : set S) {x : S} :
+  x ∈ s.centralizer ↔ ∀ a ∈ s, commute a x :=
+iff.rfl
+
+protected theorem set.mem_centralizer_iff_subset (s : set S) {x : S} :
+  x ∈ s.centralizer ↔ s ⊆ centralizer x :=
+by simp only [set.centralizer, centralizer, set.mem_set_of_eq, @commute.symm_iff _ _ x]; refl
+
+protected theorem set.centralizer_eq (s : set S) :
+  s.centralizer =  ⋂ a ∈ s, centralizer a :=
+set.ext $ assume x,
+by simp only [s.mem_centralizer, set.mem_bInter_iff, mem_centralizer]
+
+protected theorem set.centralizer_decreasing {s t : set S} (h : s ⊆ t) :
+  t.centralizer ⊆ s.centralizer :=
+by simp only [set.centralizer_eq, set.bInter_subset_bInter_left h]
+
+end defs
+
+namespace commute
+
+section semigroup
+
+variables {S : Type*} [semigroup S]
+
+/-- If `a` commutes with both `b` and `c`, then it commutes with their product. -/
+protected theorem mul {a b c : S} (hab : commute a b) (hac : commute a c) : commute a (b * c) :=
+by dunfold commute; assoc_rw [hab.eq, hac.eq]
+
+/-- If both `a` and `b` commute with `c`, then their product commutes with `c`. -/
+theorem mul_left {a b c : S}
+  (hac : commute a c) (hbc : commute b c) : commute (a * b) c :=
+(hac.symm.mul hbc.symm).symm
+
+end semigroup
+
+protected theorem all {S : Type*} [comm_semigroup S] (a b : S) : commute a b := mul_comm a b
+
+section monoid
+
+variables {M : Type*} [monoid M]
+
+protected theorem one (a : M) : commute a 1 :=
+by rw [commute, one_mul, mul_one]
+
+theorem one_left (a : M) : commute 1 a := (commute.one a).symm
+
+theorem units_inv {a : M} {u : units M} (h : commute a u) : commute a ↑u⁻¹ :=
+calc a * ↑u⁻¹ = ↑u⁻¹ * u * a * ↑u⁻¹ : by rw [units.inv_mul, one_mul]
+          ... = ↑u⁻¹ * a * u * ↑u⁻¹ : by assoc_rw [h.eq]
+          ... = ↑u⁻¹ * a            : units.mul_inv_cancel_right _ u
+
+protected theorem map {N : Type*} [monoid N] (f : M →* N) {a b : M} (h : commute a b) :
+  commute (f a) (f b) :=
+by rw [commute, ← f.map_mul, h.eq, f.map_mul]
+
+end monoid
+
+section group
+
+variables {G : Type*} [group G]
+
+protected theorem inv {a b : G} (h : commute a b) : commute a b⁻¹ :=
+@units_inv G _ a (to_units G b) h
+
+theorem inv_left {a b : G} (h : commute a b) : commute a⁻¹ b :=
+h.symm.inv.symm
+
+theorem inv_inv {a b : G} (hab : commute a b) : commute a⁻¹ b⁻¹ :=
```

Comment:
This could be an `iff` with `@[simp]`

Reply:
I can add `inv_inv_iff` (and same for `inv_left` and `inv_right`) but I'd leave this one, because now it's possible to write `(h : commute a b).inv_inv.pow_right 5`, and it's shorter than `commute.inv_inv_iff.2 h`.

Reply:
The `iff` with `simp` seems useful, so please add it. But I agree that you should also keep this one.

Reply:
Done.